state/address.go

```diff
@@ -97,39 +102,92 @@ type apiHostPortsDoc struct {
 
 // SetAPIHostPorts sets the addresses of the API server instances.
 // Each server is represented by one element in the top level slice.
-func (st *State) SetAPIHostPorts(netHostsPorts [][]network.HostPort) error {
+func (st *State) SetAPIHostPorts(newHostPorts [][]network.HostPort) error {
 	controllers, closer := st.db().GetCollection(controllersC)
 	defer closer()
-	doc := apiHostPortsDoc{
-		APIHostPorts: fromNetworkHostsPorts(netHostsPorts),
-	}
+
+	var newHostPortsForAgents [][]network.HostPort
+
 	buildTxn := func(attempt int) ([]txn.Op, error) {
-		var existingDoc apiHostPortsDoc
-		err := controllers.Find(bson.D{{"_id", apiHostPortsKey}}).One(&existingDoc)
+		var ops []txn.Op
+
+		// Retrieve the current API addresses document.
+		var existingAllAddrDoc apiHostPortsDoc
+		err := controllers.Find(bson.D{{"_id", apiHostPortsKey}}).One(&existingAllAddrDoc)
 		if err != nil {
 			return nil, err
 		}
-		op := txn.Op{
-			C:  controllersC,
-			Id: apiHostPortsKey,
-			Assert: bson.D{{
-				"txn-revno", existingDoc.TxnRevno,
-			}},
+		hostPorts := networkHostsPorts(existingAllAddrDoc.APIHostPorts)
+
+		// If the stored addresses are different from the input collection, queue an update operation.
+		if !hostsPortsEqual(newHostPorts, hostPorts) {
+			ops = append(ops, txn.Op{
+				C:  controllersC,
+				Id: apiHostPortsKey,
+				Assert: bson.D{{
+					"txn-revno", existingAllAddrDoc.TxnRevno,
+				}},
+				Update: bson.D{{
+					"$set", bson.D{{"apihostports", fromNetworkHostsPorts(newHostPorts)}},
+				}},
+			})
+		}
+
+		// Filter the collection of API addresses based on the configured management
+		// space for the controller.
+		// If there is no space configured, or if the one of the slices is filtered down
+		// to zero elements, just use the unfiltered slice for safety.
+		config, err := st.ControllerConfig()
+		if err != nil {
+			return nil, err
 		}
-		hostPorts := networkHostsPorts(existingDoc.APIHostPorts)
-		if !hostsPortsEqual(netHostsPorts, hostPorts) {
-			op.Update = bson.D{{
-				"$set", bson.D{{"apihostports", doc.APIHostPorts}},
-			}}
+		if mgmtSpace := config.JujuManagementSpace(); mgmtSpace != "" {
+			newHostPortsForAgents = make([][]network.HostPort, len(newHostPorts))
+			sp := network.SpaceName(mgmtSpace)
+			for i := range newHostPorts {
+				if filtered, ok := network.SelectHostsPortBySpaces(newHostPorts[i], sp); ok {
+					newHostPortsForAgents[i] = filtered
+				} else {
+					newHostPortsForAgents[i] = newHostPorts[i]
+				}
+			}
 		} else {
+			newHostPortsForAgents = newHostPorts
+		}
+
+		// Retrieve the current API addresses for agents document.
+		var existingAgentAddrDoc apiHostPortsDoc
+		err = controllers.Find(bson.D{{"_id", apiHostPortsForAgentsKey}}).One(&existingAgentAddrDoc)
+		if err != nil {
+			return nil, err
```

Comment:
I realize these didn't, but we should move functions to generally use:
return nil, errors.Trace(err)